src/middleware/devtools.ts

```diff
@@ -1,167 +1,325 @@
+  , unsubscribe: () => void
+  , send: (action: unknown, state: unknown) => void
+  , init: (state: unknown) => void
+  , error: (payload: unknown) => void
+  }
+
+interface DevtoolsWindow
+  { __REDUX_DEVTOOLS_EXTENSION__?:
+      { connect:
+          (options?: U.Exclude<EDevtoolsOptions, string>) => EDevtoolsExtension
       }
+  }
+
+
+
+// ============================================================================
+// Implementation
+
+type EState = { __isState: true }
+type EStore = Store<EState>
+type EDevtools = 
+  ( storeInitializer:
+      ( set: EDevtoolsStore['setState']
+      , get: EStore['getState']
+      , store: EStore
+      ) =>
+        EState
+  , options?: EDevtoolsOptions
+  ) =>
+    StoreInitializer<EState, EStore & EDevtoolsStore>
+
+
+type EDevtoolsOptions =
+  | EDevtoolsStoreName
+  | O.Overwrite<
+      U.Exclude<DevtoolsOptions, string>,
+      { name: EDevtoolsStoreName }
+    >
+type EDevtoolsStoreName =
+  undefined | (string & { __isDevtoolsStoreName: true })
+
+interface EDevtoolsStore
+  { setState:
+    (...a:
+      [...a: F.Arguments<Store<EState>['setState']>
+      , name?: EAction
+      ]
+    ) =>
+      F.Call<Store<EState>['setState']>
+  , devtools?: EDevtoolsExtension
+  , isReduxLike?: boolean
+  }
+
+type EAction = EActionType | { type: EActionType }
+type EActionType =
+  (string & { __isActionType: true }) | undefined
+
+interface EDevtoolsExtension
+  { prefix: EDevtoolsPrefix
+  , init: (state: EState | EStateFromDevtools) => void
+  , send: (..._:
+      | [action: EActionToSend, state: EState]
+      | [_: null, liftedState: EStateLifted]
+    ) => void
+  , subscribe: (listener: (message: EDevtoolsMessage) => void) => () => void
+  }
+
+type EDevtoolsMessage = 
+  | { type: "ACTION"
+    , payload: StringifiedJson<EAction> | UnknownObject
     }
-    api.setState = namedSet
-    const initialState = fn(namedSet, get, api)
-    if (!api.devtools) {
-      const savedSetState = api.setState
-      api.setState = <
-        K1 extends keyof S = keyof S,
-        K2 extends keyof S = K1,
-        K3 extends keyof S = K2,
-        K4 extends keyof S = K3
-      >(
-        state: PartialState<S, K1, K2, K3, K4>,
-        replace?: boolean
-      ) => {
-        const newState = api.getState()
-        if (state !== newState) {
-          savedSetState(state, replace)
-          if (state !== (newState as any)[DEVTOOLS] && api.devtools) {
-            api.devtools.send(api.devtools.prefix + 'setState', api.getState())
-          }
-        }
-      }
-      options = typeof options === 'string' ? { name: options } : options
-      const connection = (api.devtools = extension.connect({ ...options }))
-      connection.prefix = options?.name ? `${options.name} > ` : ''
-      connection.subscribe((message: any) => {
-        if (message.type === 'ACTION' && message.payload) {
-          try {
-            api.setState(JSON.parse(message.payload))
```

Comment:
Here `JSON.parse(message.payload)` would give the action to be dispatched (afaict) and not the state, so this would be buggy, fixed in rewrite

Reply:
I think it was intentional. We put `state` in the action object. cc @Aslemammad 
We could change the behavior, but I want to have it in a different PR (possibly in advance.)

Reply:
Ah yes you're write we're sending the state as action:
https://github.com/pmndrs/zustand/blob/c8ae965435beb2cfb5a8bb8106e4e81712e9de25/src/middleware/devtools.ts#L106

In that case I'll keep the behavior same at least in this PR and open a another PR for changing the behavior (will reread the existing code to see if what I'm thinking is possible)

Reply:
Actually no, this does not change the behavior afaict, [`send` has signature `(action, state) => void`](https://github.com/zalmoxisus/redux-devtools-extension/blob/master/docs/API/Methods.md#returns), so we're sending `prefix + 'setState'` as the action and not the state itself. But still I'll open a seperate PR for refactoring this.